### PR TITLE
Feature/3516 user consent switch automatic result sharing

### PR DESCRIPTION
**Pull request by puneetmahali1**
## Description

- This PR is related to the submission flow based on consent Switch. 
- Modify Your Test Result screen based on switch state.
- Newly Build Your Consent screen.
- Handel the switch state when user gave the consent or not.

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-3516

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Models/Exposure/__tests__/WarnOthersReminderTests.swift [hunk 1]
```diff
@@ -53,4 +63,28 @@ class WarnOthersReminderTests: XCTestCase {
...
+	
+	
+	func testWarnOthers_SubmissionConsentGiven() throws {
+		
+		let store = MockTestStore()
+		store.isSubmissionConsentGiven = true
+		
+		let warnOthersReminder = WarnOthersReminder(store: store)
+		
+		XCTAssertTrue(warnOthersReminder.isSubmissionConsentGiven, "Submission consent given should be true")
```
Comment (marcussc) [inline src/xcode/ENA/ENA/Source/Models/Exposure/__tests__/WarnOthersReminderTests.swift:75]: I like the description. Will make understanding the errors easier on the CI!

**Comment by marcussc** [comment]
VoiceOver seems to have a little hiccup. When reading the country's in the gray block it start with "coma Belgium …"

File: src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewConsentCell.swift [hunk 2]
```diff
@@ -0,0 +1,184 @@
+//
+// 🦠 Corona-Warn-App
+//
+
+import UIKit
+
+class DynamicTableViewConsentCell: UITableViewCell {
```
Comment (marcussc) [inline src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewConsentCell.swift:7]: Not sure if it is this cell but when I swipe around on the screen "Ihr Einverständnis" i get some Warnings in Xcode about UnsatisfiableConstraints.

```
2020-11-24 10:18:59.262185+0100 ENA[2500:3157540] [LayoutConstraints] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x280029360 UITableViewCellContentView:0x110e5ffd0.height == 50   (active)>",
    "<NSLayoutConstraint:0x2800d4fa0 'UIView-Encapsulated-Layout-Height' UITableViewCellContentView:0x110e5ffd0.height == 20   (active)>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x280029360 UITableViewCellContentView:0x110e5ffd0.height == 50   (active)>```

**Review by marcussc — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewConsentCell.swift (hunk 2 above)
Comment (puneetmahali1) [reply src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewConsentCell.swift:7]: Yes, we know that about this breaking constraint. We digging a lot for this moment and not find the solution right now that's why I have created a bug ticket for that. Here is bug ticket link: 
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-3938

File: src/xcode/ENA/ENA/Source/Coordinator.swift [hunk 3]
```diff
@@ -189,7 +190,7 @@ extension Coordinator: HomeViewControllerDelegate {
 		// when .start() is called. The coordinator is then bound to the lifecycle of this navigation controller
 		// which is managed by UIKit.
 		let coordinator = ExposureSubmissionCoordinator(
-			warnOthersReminder: warnOthersReminder,
+			warnOthersReminder: WarnOthersReminder(store: self.store),
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Coordinator.swift:193]:
```suggestion
			warnOthersReminder: WarnOthersReminder(store: store),
```

File: src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewConsentCell.swift [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		}
+		consentStackView.setCustomSpacing(16, after: seperatorView1)
+		consentStackView.setCustomSpacing(16, after: flagCountriesLabel)
+		consentStackView.setNeedsUpdateConstraints()
+	}
+	
+	
+	private func setupConstraints() {
+		let marginGuide = contentView.layoutMarginsGuide
+
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewConsentCell.swift:123]: remove one empty line

File: src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewConsentCell.swift [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+extension NSTextAttachment {
+	func setImageHeight(height: CGFloat) {
+		guard let image = image else { return }
+		let ratio = image.size.width / image.size.height
+
+		bounds = CGRect(x: bounds.origin.x, y: bounds.origin.y, width: ratio * height, height: height)
+	}
+}
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewConsentCell.swift:184]: Do we have an extra file for NSTestAttachments yet? If so, move it over there overwise create a new file, so others can finde this more easy

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewController.swift [hunk 6]
```diff
@@ -0,0 +1,58 @@
+//
+// 🦠 Corona-Warn-App
+//
+
+import UIKit
+import Combine
+
+class ExposureSubmissionTestResultConsentViewController: DynamicTableViewController {
+
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewController.swift:9]: remove one blank line please

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewController.swift [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	
+	init(
+		supportedCountries: [Country],
+		exposureSubmissionService: ExposureSubmissionService
+	) {
+				
+		self.viewModel = ExposureSubmissionTestResultConsentViewModel(supportedCountries: supportedCountries, exposureSubmissionService: exposureSubmissionService)
+		super.init(nibName: nil, bundle: nil)
+	}
+
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewController.swift:21]: why use supportedCountries and exposureSubmissionService as parameters to only create a viewModel here?
You can create the viewmodel outside and pass into the init method

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewController.swift [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+			forCellReuseIdentifier: CustomCellReuseIdentifiers.consentCell.rawValue
+		)
+	}
+}
+
+extension ExposureSubmissionTestResultConsentViewController {
+	enum CustomCellReuseIdentifiers: String, TableViewCellReuseIdentifiers {
+		case consentCell
+	}
+}
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewController.swift:58]: This should be part of the view model?

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+		
+		self.exposureSubmissionService = exposureSubmissionService
+		
+		self.supportedCountries = supportedCountries.sorted { $0.localizedName.localizedCompare($1.localizedName) == .orderedAscending }
+		
+		self.exposureSubmissionService.isSubmissionConsentGivenPublisher.sink { isSubmissionConsentGiven in
+			self.consentSwitch.isOn = isSubmissionConsentGiven
+		}.store(in: &cancellables)
+	}
+
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewModel.swift:25]: blank line, please remove one

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+							text: AppStrings.AutomaticSharingConsent.switchTitle,
+							color: nil,
+							accessibilityIdentifier: nil,
+							accessibilityTraits: .header,
+							configure: { _, cell, _ in
+								cell.accessoryView = self.consentSwitch
+								self.consentSwitch.onTintColor = .enaColor(for: .tint)
+								self.consentSwitch.addTarget(self, action: #selector(self.consentStateChanged), for: .valueChanged)
+							}
+						),
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewModel.swift:62]: capture self inside this closure might leak?

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+						.custom(
+							withIdentifier: ExposureSubmissionTestResultConsentViewController.CustomCellReuseIdentifiers.consentCell,
+							action: .none,
+							accessoryAction: .none) { _, cell, _ in
+								if let consentCell = cell as? DynamicTableViewConsentCell {
+									consentCell.configure(
+										subTitleLabel: NSMutableAttributedString(string: AppStrings.AutomaticSharingConsent.consentSubTitle),
+										descriptionPart1Label: NSMutableAttributedString(string: AppStrings.AutomaticSharingConsent.consentDescriptionPart1),
+										descriptionPart2Label: NSMutableAttributedString(string: AppStrings.AutomaticSharingConsent.consentDescriptionPart2),
+										countries: self.supportedCountries,
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewModel.swift:73]: self captured inside this closure might leak

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift [hunk 12]
```diff
@@ -6,47 +6,49 @@ import UIKit
...
+	
 	init(
 		warnOthersReminder: WarnOthersRemindable,
 		testResult: TestResult,
 		exposureSubmissionService: ExposureSubmissionService,
 		onContinueWithSymptomsFlowButtonTap: @escaping (@escaping (Bool) -> Void) -> Void,
 		onContinueWithoutSymptomsFlowButtonTap: @escaping (@escaping (Bool) -> Void) -> Void,
-		onTestDeleted: @escaping () -> Void
+		onTestDeleted: @escaping () -> Void,
+		onSubmissionConsentButtonTap: @escaping (@escaping (Bool) -> Void) -> Void
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift:19]: escaping closure use escaping closures as a parameter?
This looks strange to me and might indicate a problem. We should discuss if there is a better solution to do this.

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift [hunk 13]
```diff
@@ -165,206 +176,243 @@ class ExposureSubmissionTestResultViewModel {
...
-					hairline: .topAttached
+			.section(
+				separators: .all,
+				cells: [
+					
+					.icon(
+						UIImage(imageLiteralResourceName: "Icons_Grey_Warnen"),
+						text: .string(self.submissionConsentLabel),
+						action: .execute { _, cell in
+							self.onSubmissionConsentButtonTap { isLoading in
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift:352]: self inside a closure might leak

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift [hunk 14]
```diff
@@ -165,206 +176,243 @@ class ExposureSubmissionTestResultViewModel {
...
+								let activityIndicatorView = UIActivityIndicatorView(frame: CGRect(x: 0, y: 0, width: 20, height: 20))
+								activityIndicatorView.startAnimating()
+								cell?.accessoryView = isLoading ? activityIndicatorView : nil
+								cell?.isUserInteractionEnabled = !isLoading
+								self.navigationFooterItem.isPrimaryButtonEnabled = !isLoading
+								self.navigationFooterItem.isSecondaryButtonEnabled = !isLoading
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift:358]: self in closure might leak

**Review by KaiTeuber — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewModel.swift (hunk 9 above)
Comment (puneetmahali1) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewModel.swift:25]: Done.

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewController.swift (hunk 6 above)
Comment (puneetmahali1) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewController.swift:9]: Thanks, Done.

File: src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewConsentCell.swift (hunk 4 above)
Comment (puneetmahali1) [reply src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewConsentCell.swift]: Thanks, Done.

File: src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewConsentCell.swift (hunk 5 above)
Comment (puneetmahali1) [reply src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewConsentCell.swift:183]: Done. Move into the extension classes folder.

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewController.swift (hunk 7 above)
Comment (puneetmahali1) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewController.swift:20]: Done.

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewModel.swift (hunk 10 above)
Comment (puneetmahali1) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewModel.swift:61]: Put the weak reference[weak]. Solved.

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewModel.swift (hunk 11 above)
Comment (puneetmahali1) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewModel.swift:72]: Put the weak reference[weak]. Solved.

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift (hunk 13 above)
Comment (puneetmahali1) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift:352]: Put the weak reference[weak]. Solved.

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewController.swift (hunk 8 above)
Comment (puneetmahali1) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultConsentViewController.swift]: Thanks kai, We did it.

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift (hunk 12 above)
Comment (puneetmahali1) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift:19]: As we discussed and agreed, we will think it later on.

**Review by KaiTeuber — APPROVED** [review]

**Review by marcussc — APPROVED** [review]

**Review by SwissMountainFrog — APPROVED** [review]

**Review by tcfos — APPROVED** [review]

**Pull request merged**